rfcs/20240311-gather-scatter-batching-dims.md

```diff
@@ -0,0 +1,427 @@
+* (C16) `size(operand_batching_dims) == size(start_indices_batching_dims)`.
+* (C17) `dim(operand, operand_batching_dims...) = dim(start_indices, start_indices_batching_dims...)`.
+* (C18) `is_unique(start_index_map + operand_batching_dims)`.
+* (C19) `0 <= start_index_map < rank(operand)`.
+* (C20) `size(slice_sizes) = rank(operand)`.
+* (C21) `0 <= slice_sizes <= shape(operand)`.
+* (C22) `shape(result) = combine(batch_dim_sizes, offset_dim_sizes)` where:
+  * `batch_dim_sizes = shape(start_indices)` except that the dimension size
+    of `start_indices` corresponding to `index_vector_dim` is not included.
+  * `offset_dim_sizes = shape(slice_sizes)` except that the dimension sizes
+    in `slice_sizes` corresponding to `collapsed_slice_dims` and
+    `operand_batching_dims` are not included.
+  * `combine` puts `batch_dim_sizes` at axes corresponding to `batch_dims` and
+   `offset_dim_sizes` at axes corresponding to `offset_dims`.
+* (C23) `element_type(operand) = element_type(result)`.
+
+#### Examples
+
+```mlir
+// %operand: [
+//            [
+//             [[1, 2], [3, 4], [5, 6], [7, 8]],
+//             [[9, 10],[11, 12], [13, 14], [15, 16]],
+//             [[17, 18], [19, 20], [21, 22], [23, 24]]
+//            ],
+//            [
+//             [[25, 26], [27, 28], [29, 30], [31, 32]],
+//             [[33, 34], [35, 36], [37, 38], [39, 40]],
+//             [[41, 42], [43, 44], [45, 46], [47, 49]]
+//            ]
+//           ]
+// %start_indices: [
+//                  [
+//                   [[0, 0], [1, 0], [2, 1]],
+//                   [[0, 1], [1, 1], [0, 9]]
+//                  ],
+//                  [
+//                   [[0, 0], [2, 1], [2, 2]],
+//                   [[1, 2], [0, 1], [1, 0]]
+//                  ]
+//                 ]
+%result = "stablehlo.gather"(%operand, %start_indices) {
+  dimension_numbers = #stablehlo.gather<
+    offset_dims = [3, 4],
+    collapsed_slice_dims = [1],
+    operand_batching_dims = [0],
+    start_indices_batching_dims = [1],
+    start_index_map = [2, 1],
+    index_vector_dim = 3>,
+  slice_sizes = array<i64: 1, 1, 2, 2>,
+  indices_are_sorted = false
+} : (tensor<2x3x4x2xi32>, tensor<2x2x3x2xi64>) -> tensor<2x2x3x2x2xi32>
+// %result: [
+//           [
+//            [
+//             [[1, 2], [3, 4]],
+//             [[3, 4], [5, 6]],
+//             [[13, 14], [15, 16]]
+//            ],
+//            [
+//             [[33, 34], [35, 36]],
+//             [[35, 36], [37, 38]],
+//             [[41, 42], [43, 44]]
+//            ]
+//           ],
+//           [
+//            [
+//             [[1, 2], [3, 4]],
+//             [[13, 14], [15, 16]],
+//             [[21, 22], [23, 24]]
+//            ],
+//            [
+//             [[43, 44], [45, 46]],
+//             [[33, 34], [35, 36]],
+//             [[27, 28], [29, 30]]
+//            ]
+//           ]
+//          ]
+```
+
+&nbsp;[More Examples](../stablehlo/tests/interpret/gather.mlir)
+
+### scatter
```

Comment:
In RFC, may you add 1-2 paragraph(s) to describe the *changed* part in the proposal (so that it will be easier to understand the RFC)?

For example:

    This RFC proposed adding `input_batching_dimensions` and `scatter_indices_batching_dimensions` attributes
    to `stablehlo.scatter`.  `input_batching_dimensions` refers to the dimensions of the `inputs` that are treated
    as batch.  `scatter_indices_batching_dimensions` refers to the dimensions of the `scatter_indices` that are treated
    as batch.  The corresponding dimension sizes must be equal.  The semantics is equivalent to concatenating
    the outputs of the scatter with each slices of `input` and `scatter_indices`.

And then follow by the full proposed new wording for completeness (as you have done).

Reply:
Done.